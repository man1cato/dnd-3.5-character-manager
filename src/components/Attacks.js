import React from 'react';

const Attacks = (props) => (
    <div>
        <div className="row">Base Attack Bonus: {props.bab}</div>
        <div className=" grid grid--6col">
            <h4 className="grid__left-col">Attack Type</h4>
            <h4>Base</h4>
            <div></div>
            <h4>Temp Mod</h4>
            <div></div>
            <h4>Total</h4>

            <div className="grid__left-col">Melee</div>
            <div>{props.melee.base}</div>
            <div>+</div>
            <input
                type="text"
                name="melee"
                value={props.melee.mod}
                onChange={props.onInputChange}
            />
            <div>=</div>
            <div>{props.melee.total}</div>

            <div className="grid__left-col">Ranged</div>
            <div>{props.ranged.base}</div>
            <div>+</div>
            <input
                type="text"
                name="ranged"
                value={props.ranged.mod}
                onChange={props.onInputChange}
            />
            <div>=</div>
            <div>{props.ranged.total}</div>

            <div className="grid__left-col">Grapple</div>
            <div>{props.grapple.base}</div>
            <div>+</div>
            <input
                type="text"
                name="grapple"
                value={props.grapple.mod}
                onChange={props.onInputChange}
            />
            <div>=</div>
            <div>{props.grapple.total}</div>
        </div>
    </div>
)

export default Attacks