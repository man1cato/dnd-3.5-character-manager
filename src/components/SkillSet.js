import React from 'react'
import { connect } from 'react-redux'
import SkillModal from './SkillModal'

export class SkillSet extends React.Component{
	state = {
		selected: undefined
	}  

	handlePick = (e) => {
		const id = e.target.id
		const selected = this.props.skills[id]
		this.setState({selected})
	}

	handleCloseModal = () => {
		this.setState({selected: undefined})
	}
	
	render () {
		const skillCount = this.props.skillSet.length
		const col1 = this.props.skillSet.slice(0, Math.ceil(skillCount / 2))
		const col2 = this.props.skillSet.slice(Math.ceil(skillCount / 2))
		
		return (
			<div className="grid--skills">
				<h5 className="grid--skills__name">Skill</h5>
				{col1.map((skill, i) => (        
					<button 
						className="grid__col1 grid--skills__name button--link" 
						id={skill.id}
						key={i}
						onClick={this.handlePick}
					>
						{this.props.skills[skill.id].name}
					</button>
				))}

				<h5>Ranks</h5>
				{col1.map((skill, i) => (
					<div className="grid__col2" key={i}>{skill.ranks}</div>
				))}
				
				<h5 className="grid--skills__name">Skill</h5>
				{col2.map((skill, i) => (
					<button 
						className="grid__col3 grid--skills__name button--link" 
						id={skill.id}
						key={i}
						onClick={this.handlePick}
					>
						{this.props.skills[skill.id].name}
					</button>
				))}
				
				<h5>Ranks</h5>        
				{col2.map((skill, i) => (
					<div className="grid__col4" key={i}>{skill.ranks}</div>
				))}

				<SkillModal 
					selected={this.state.selected}
					handleCloseModal={this.handleCloseModal}
				/>
			</div>
		)
	}
}


const mapStateToProps = (state) => ({
  	skills: state.skills
})

export default connect(mapStateToProps)(SkillSet)