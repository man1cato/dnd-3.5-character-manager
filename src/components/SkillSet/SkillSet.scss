@import '../../styles/base/settings';


.SkillSet {
   display: grid;
   grid-template-columns: 3fr 1fr 3fr 1fr;
   grid-auto-flow: dense;
   grid-row-gap: $xs-size; 
   grid-column-gap: $s-size;
   justify-items: center;
   text-align: center; 
   align-items: start;
   > input {
      max-width: 50px;
      text-align: center;
   }  
}

.SkillSet__name-header {
  justify-self: start;
  text-align: left;
}

.SkillSet__col3 {
   grid-column: 3;
   justify-self: self-start;
}