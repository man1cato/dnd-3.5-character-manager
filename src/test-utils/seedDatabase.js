import _ from 'lodash'
import profile from './fixtures/profile'
import database from '../firebase/firebase'


export const characterOne = {
   ..._.omit(profile, ['id'])
}

export const characterTwo = {
   name: 'Korg',
   gender: 'Male',
   race: 'recoDD3gAaQXU3A56',
   jobClass: 'recbxccm8pp3fTmQw',
   alignment: 'Lawful Evil',
   level: 5,
   iconUrl: 'images/orc.svg'
}

export const characterThree = {
   name: 'Rylia',
   gender: 'Female',
   race: 'recFEpQ9zuD9Yqu7l',
   jobClass: 'recQrfCKtksnnSsaZ',
   alignment: 'Chaotic Neutral',
   level: 3,
   iconUrl: 'images/halfling.svg'
}

export const userOne = {
   uid: 'userOne',
   characterIds: ['characterOne', 'characterTwo']
}

export const userTwo = {
   uid: 'userTwo',
   characterIds: ['characterThree']
}


export default async () => {
   await database.ref('users').remove()
   await database.ref(`users/${userOne.uid}/profiles/${userOne.characterIds[0]}`).set(characterOne)
   await database.ref(`users/${userOne.uid}/profiles/${userOne.characterIds[1]}`).set(characterTwo)
   await database.ref(`users/${userTwo.uid}/profiles/${userTwo.characterIds[0]}`).set(characterThree)
}