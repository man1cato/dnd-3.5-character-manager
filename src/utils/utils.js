import _ from 'lodash'


export const convertTextToArray = (string) => _.compact(string.split(/\n/))

export const findItemById = (storeItemArray, itemId) => storeItemArray.find((item) => item.id === itemId)

export const apiObjectToArray = (obj) => _.orderBy(Object.keys(obj).map((id) => ({
	id,
	...obj[id]
})), ['name'], ['asc'])