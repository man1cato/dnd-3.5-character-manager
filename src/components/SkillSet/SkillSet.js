import React, { Fragment, useState } from 'react'
import { connect } from 'react-redux'
import _ from 'lodash'

import SkillModal from '../Modals/SkillModal'
import './SkillSet.scss'


export const SkillSet = props => {
	const[clickedSkill, setClickedSkill] = useState(null)

	const skillCount = props.skillSet.length
	const col1 = props.skillSet.slice(0, Math.ceil(skillCount / 2))
	const col2 = props.skillSet.slice(Math.ceil(skillCount / 2))
	
	return (
		<div className="SkillSet">
			<h5 className="SkillSet__name-header">Skill</h5>
			<h5>Ranks</h5>
			<h5 className="SkillSet__name-header">Skill</h5>
			<h5>Ranks</h5>        

			{col1.map(skill => ( 
				<Fragment key={skill.id}>       
					<button 
						className="grid__col1 button--link" 
						id={skill.id}
						type='button'
						onClick={() => setClickedSkill(props.skills[skill.id])}
					>
						{props.skills[skill.id].name}
					</button>
					<div className="grid__col2">{skill.ranks}</div>
				</Fragment>
			))}

			{col2.map(skill => (
				<Fragment key={skill.id}>
					<button 
						className="SkillSet__col3 button--link" 
						id={skill.id}
						type='button'
						onClick={() => setClickedSkill(props.skills[skill.id])}
					>
						{props.skills[skill.id].name}
					</button>
					<div className="grid__col4">{skill.ranks}</div>
				</Fragment>
			))}
			
			<SkillModal 
				clickedSkill={clickedSkill}
				handleCloseModal={() => setClickedSkill(null)}
			/>
		</div>
	)
}


const mapStateToProps = state => ({
  	skills: state.skills
})

export default connect(mapStateToProps)(SkillSet)