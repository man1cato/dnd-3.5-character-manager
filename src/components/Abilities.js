import React, { Fragment } from 'react'
import _ from 'lodash'
import { calcAbilityMod } from '../utils/utils'


const Abilities = ({ abilities, handleChange }) => (
	<div className="grid--abilities">
		<h5 className="grid__col1">Ability</h5>
		<h5 className="grid__col2">Score</h5>
		<h5 className="grid__col3">Temp Score</h5>
		<h5 className="grid__col4"> Mod</h5>

		{_.map(_.keys(abilities), (ability, i) => (
			<Fragment key={i}>
				<div className="grid__col1">{ability.toUpperCase()}</div>
				<div className="grid__col2">{abilities[ability].score}</div>
				<input
					className="grid__col3"
					id={ability}
					name="abilities"
					type="number"
					value={abilities[ability].tempScore}
					onChange={handleChange}
				/>
				<div className="grid__col4">
					{!!abilities[ability].tempScore ? calcAbilityMod(abilities[ability].tempScore) : calcAbilityMod(abilities[ability].score)}
				</div>
			</Fragment>
		))}
	</div>
)


export default Abilities