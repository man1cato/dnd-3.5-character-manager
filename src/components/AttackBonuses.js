import React, { useState, useEffect, Fragment } from 'react'
import _ from 'lodash'
import update from 'immutability-helper'

import { calcBaseMeleeBonus, calcBaseRangedBonus, calcBaseGrappleBonus, convertInputValue } from '../utils/utils'


const AttackBonuses = ({ profileAttackBonusMods, abilities, baseAttackBonuses, size, handleUpdate }) => {  
    const attackBonusBases = {
        melee: _.map(baseAttackBonuses, (bab) => calcBaseMeleeBonus(bab, abilities.str.score, size)),
        ranged: _.map(baseAttackBonuses, (bab) => calcBaseRangedBonus(bab, abilities.dex.score, size)),
        grapple: _.map(baseAttackBonuses, (bab) => calcBaseGrappleBonus(bab, abilities.str.score, size))
    }
    const [attackBonusMods, setAttackBonusMods] = useState(_.mapValues(profileAttackBonusMods, (val) => val || '')) 
    const [attackBonusTotals, setAttackBonusTotals] = useState(
        _.mapValues(attackBonusMods, (mod, key) => _.map(attackBonusBases[key], (val) => Number(mod) + val ))
    ) 
    
    useEffect(() => {
        setAttackBonusTotals(_.mapValues(attackBonusMods, (mod, key) => _.map(attackBonusBases[key], (val) => Number(mod) + val)))
        handleUpdate({ attackBonusMods })
    }, [attackBonusMods])

    return (    
        <div className="grid--6col">
            <h5 className="grid__col1">Type</h5>
            <h5>Base</h5>
            <div></div>
            <h5>Mod</h5>
            <div></div>
            <h5>Total</h5>

            {_.map(['melee', 'ranged', 'grapple'], (key) => (
                <Fragment key={key}>
                    <div className="grid__col1">{_.startCase(key)}</div>
                    <div>{attackBonusBases[key].join(' / ')}</div>
                    <div>+</div>
                    <input
                        type="number"
                        id={key}
                        value={attackBonusMods[key]}
                        onChange={(e) => {
                            setAttackBonusMods(
                                update(attackBonusMods, {
                                    [key]: { $set: convertInputValue(e.target.value) }
                                })
                            )
                        }}
                    />
                    <div>=</div>
                    <div>{attackBonusTotals[key].join(' / ')}</div>
                </Fragment>
            ))}

        </div>
    )
}


export default AttackBonuses