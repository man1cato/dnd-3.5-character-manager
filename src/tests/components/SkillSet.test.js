import React from 'react'
import { render, fireEvent } from '@testing-library/react'

import { SkillSet } from '../../components/SkillSet'
import { apiData } from '../utils/utils'
import { characterOne } from '../utils/seedDatabase'


const skillSet = characterOne.skillSet
let props
beforeAll(async () => {
	const api = await apiData()
	props = {
		skillSet,
		skills: api.skills
	}
})


test('should render SkillSet with profile data', () => {
	const { container } = render(<SkillSet {...props} />)
	expect(container.firstChild).toMatchSnapshot()
})

test('should update selected in state on skill button click', async () => {
	const { getByText, queryByLabelText, findByLabelText } = render(<SkillSet {...props} />)
	const labelMatch = 'Selected Skill'
	
	expect(queryByLabelText(labelMatch)).toBe(null)

	const textMatch = props.skills[skillSet[0].id].name
	fireEvent.click(getByText(textMatch))

	const modal = await findByLabelText(labelMatch)
	expect(modal).toBeDefined()
})