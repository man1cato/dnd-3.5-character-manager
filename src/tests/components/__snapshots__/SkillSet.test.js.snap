// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render SkillSet with profile data 1`] = `
<div
  className="grid--skills"
>
  <h5
    className="grid--skills__name"
  >
    Skill
  </h5>
  <button
    className="grid__col1 grid--skills__name button--link"
    id="recfbplSsRmzNs68J"
    key="0"
    onClick={[Function]}
  >
    Appraise
  </button>
  <button
    className="grid__col1 grid--skills__name button--link"
    id="recoJ7dkAWtHAkj3v"
    key="1"
    onClick={[Function]}
  >
    Balance
  </button>
  <h5>
    Ranks
  </h5>
  <div
    className="grid__col2"
    key="0"
  >
    4
  </div>
  <div
    className="grid__col2"
    key="1"
  >
    3
  </div>
  <h5
    className="grid--skills__name"
  >
    Skill
  </h5>
  <button
    className="grid__col3 grid--skills__name button--link"
    id="recHCeKVn3CB0HcNA"
    key="0"
    onClick={[Function]}
  >
    Climb
  </button>
  <button
    className="grid__col3 grid--skills__name button--link"
    id="recs2EXcqiCIfwO0w"
    key="1"
    onClick={[Function]}
  >
    Concentration
  </button>
  <h5>
    Ranks
  </h5>
  <div
    className="grid__col4"
    key="0"
  >
    5
  </div>
  <div
    className="grid__col4"
    key="1"
  >
    9
  </div>
  <SkillModal
    handleCloseModal={[Function]}
  />
</div>
`;
